bpf:trace: adjust TRACE_PAYLOAD_LEN for overlay packets

smagnani96 opened the pull request:
~Blocked by https://github.com/cilium/cilium/pull/39650.~ Good to have (non-blocking) for https://github.com/cilium/cilium/pull/37634.

As described in commit messages, when tracing overlay packets the default TRACE_PAYLOAD_LEN might not be enough. An example would be an Overlay IPv4 packet carrying an IPv6 TCP packet with Options (MSS, etc., such as a SYN packet). In that case, Hubble would fail decoding such packet due to missing bytes. Same problems would arise in case of IPv6 underlay.

For this reason:
1. move `TRACE_PAYLOAD_LEN` to `CONFIG(trace_payload_len)` for coherency with the new datapath config, and use it to trace native packets
2. introduce `CONFIG(trace_payload_len_overlay)` and use it for tracing overlay packets
3. postpone the computation of capture_length when tracing to only when the trace needs to be omitted
4. add an helper to accomplish (3) based on the current classifier flags: use (2) when CLS_FLAG_{VXLAN,GENEVE} is set, (1) otherwise.

Fixes: https://github.com/cilium/cilium/issues/39030.

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -539,8 +539,12 @@ enum {
 
+/* Minimum bytes for decoding packet headers during trace events.
+ * - Outer: 14 (ethernet) + 40 (ipv6) + 8 (udp) + 8 (geneve) + 24 (geneve v6 opts)
+ * - Inner: 14 (ethernet) + 40 (ipv6) + 8 (udp)
+ */
 #ifndef TRACE_PAYLOAD_LEN
-#define TRACE_PAYLOAD_LEN 128ULL
+#define TRACE_PAYLOAD_LEN 156ULL
```
I wonder whether we really need to bump the default size. Or can have a `TRACE_PAYLOAD_OVERLAY_LEN`, and plug it in at the right places (based on `ctx_is_overlay()` etc).

smagnani96 replied:
Thanks for the precious feedback. I tried to address it at my best, just updated the PR to achieve a similar behavior. 🙏🏼 

Closing this since outdated code, but super happy to start a new one!

julianwiedmann commented:
> Outer: 14 (ethernet) + 40 (ipv6) + 8 (udp) + 8 (geneve) + 24 (geneve v6 opts)

You might want to consider one level of VLAN header in this calculation, to be on the safe side.

smagnani96 commented:
Marking this as ready-for-review to start gathering feedback.
Will wait for the parent PR to get merged first: there shouldn't be substantial changes in between, and this PR is a good-to-have for the final Hubble PR.

smagnani96 commented:
Haven't run `ci-verifier` yet in this PR, but let me also report an interesting data.
Comparing locally kernel v6.6 N. verified instruction `THIS_BRANCH (after) - main (before)`:

![bpf_host_6](https://github.com/user-attachments/assets/2e02359c-f366-4157-be4e-a1cc520088b6)

Worst case increase +12k, Best case saved -350k. If found to be useful, I can also think of a partial backport of this PR. In that case, I believe most relevant commits would be:

* `bpf: move TRACE_PAYLOAD_LEN to node config variable`
* `bpf:classifiers: add ctx_capture_length for postponing computation`, the postponing part.

Here's the full zip for curious [insn-diff-66_log-wrt-main66_log-20250609_173029.zip](https://github.com/user-attachments/files/20657263/insn-diff-66_log-wrt-main66_log-20250609_173029.zip).

tommyp1ckles commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -387,7 +387,7 @@ cilium-agent [flags]
       --tofqdns-proxy-port int                                    Global port on which the in-agent DNS proxy should listen. Default 0 is a OS-assigned port.
       --tofqdns-proxy-response-max-delay duration                 The maximum time the DNS proxy holds an allowed DNS response before sending it along. Responses are sent as soon as the datapath is updated with the new IP information. (default 100ms)
-      --trace-payloadlen int                                      Length of payload to capture when tracing (default 128)
+      --trace-payloadlen int                                      Length of payload to capture when tracing packets. (default 128)
```
Is this auto aligned to the cache line size? should we mention that or otherwise specify that it needs to be that.

tommyp1ckles commented on bpf/lib/classifiers.h:
```diff
@@ -36,7 +36,49 @@ enum {
+ * supports_overlay_mark
+ * @obs_point: trace observation point (TRACE_{FROM,TO}_*)
+ *
+ * Returns true whether the provided observation point can observe Overlay traffic marked with the
```
nit: observe Overlay traffic -> observe overlay traffic

smagnani96 committed:
bpf: move TRACE_PAYLOAD_LEN to node config variable

With this commit, we move TRACE_PAYLOAD_LEN from being a pure macro to being
a config variable. This is purely to adapt to new BPF config and subsequent
changes to introduce a new related config variable for tracing overlay packets.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: add trace_payload_len_overlay variable for overlay packets

This commit introduces a new node config variable trace_payload_len_overlay.
This variable will be enabled in subsequent commit, so that we can push
more bytes during trace notification for overlay traffic. In fact, the
old TRACE_PAYLOAD_LEN (128B) might not be enough for Hubble to decode
overlay packets such as:

- TCPv6 with options over VXLANv4 (>=134B) -- decode error
- TCPv6 with SRv6 segments (>=136B) -- decode error
- {ICMP,UDP,TCP}v6 over (future) VXLANv6 (>=132B) -- decode error

Given that now the datapath can distinguish between native and overlay
traffic, let's make use of it.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:classifiers: refactor overlay checks into helpers functions

With this patch we simply refactors the checks to enable Overlay
classifiers during classification into specific helper functions for
code clarity and reusability (see subsequent patch).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: add compute_capture_length for postponing notification length

This commits adds an helper function `compute_capture_length` to compute the
length of the capture associated to the trace/drop notification event.
While doing this, we now postpone the computation of such value in trace/drop
notification to save computational time and verifier complexity, given
the value would not be used in case the notification should not be emitted.
With the next commit, we further extend this helper to account for
Overlay packets that require additional bytes.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf: use trace_payload_len_overlay when tracing overlay packets

This commit make use of the previously introduced trace_payload_len_overlay.
Whenever the dataplane is able to classify a packet as OVERLAY, let's
push additional bytes in the notification event to allow Hubble decoding
the packet. An example of failure before this patch is the `TCP data offset
greater than packet length` error. This error can happen with an TCPv6 SYN
(options MSS) over VXLAN: we're pushing 128B, but 144B would be needed for
correctly decoding the TCP options.

The utility function `ctx_capture_length` returns the correct capture
length for the notification event. It makes use of classifiers flags
and observation point. When provided with the monitor value (trace
notifications, as in drop notifications we use 0), the helper accounts for it:

* if `monitor` is 0, then the utility selects the default value based on
  the provided flags, either trace_payload_len or trace_payload_len_overlay.
* if monitor is the default trace_payload_len used in the codebase, then
  the helper still checks the flags and transparently adjusts to
  trace_payload_len_overlay with overlay traffic.

Normal usages should still prefer trace_payload_len in the codebase: the
dapath will transparently adjust it to trace_payload_len_overlay during
the trace notification event if/when needed.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 commented:
/test
